rfd/0078-login-rules.md

```diff
@@ -0,0 +1,504 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 78 - Login Rules
+
+## What
+
+A new feature which allows admins to filter, transform, and extend incoming
+SAML assertions and OIDC claims from their identity provider during user login
+before they are embedded in the traits of their Teleport user's certificates.
+
+## Why
+
+Teleport admins often do not have control over their company's SSO provider and
+which claims will be provided to Teleport.
+All of these claims will be embedded in each user's Teleport certificate as
+`traits`.
+
+SSO providers can include hundreds to thousands of claims which are not
+necessary or used by Teleport at all, unecessarily bloating the certificate size
+of all users.
+This can adversly impact latency and throughput of Teleport operations, and in
+an extreme case could reach a limit which would render Teleport unusable with
+that identity provider.
+Login rules will allow the Teleport admin to filter out claims which are
+unecessary before they make it into the Teleport certificate.
+
+Admins often wish to extend an incoming claim with extra values which make sense
+for their Teleport deployment.
+For example, users within the `splunk` group should also be added to the `dbs`
+group within Teleport so that they can access databases.
+Login rules will allow the Teleport admins to modify existing traits and
+add new ones to solve these problems.
+
+If Teleport admins cannot add or change the claims provided by their SSO
+provider, making simple rules like this can be extremely cumbersome or
+impossible within teleport.
+There are currently teleport users with deployments which automatically generate
+hundreds to thousands of roles to get around these limitations.
+
+## Details
+
+Example login rule spec:
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: example
+spec:
+  # priority can be used to order the evaluation of multiple login rules within
+  # a cluster. Lower priorities will be evaluated first. Login rules with the
```

Comment:
> It's common for integers to default to zero, making those rules jump to the top of the priority list by default

Where is the top? It's not clear. A priority 0 rule will be evaluated first, but its output will be completely overwritten by a rule at priority 1, the latter p1 rule gets the final say.

Reply:
If, as you said, priority is a dubious concept, should we call it "evaluation order" instead?

Whatever you want to name it, if we go with an int field then it will default to zero when not explicitly defined, which is something to consider.